gpt.lua

```diff
@@ -0,0 +1,555 @@
+  skip = 0
+
+  if current_status ~= Status.start then
+    debug_log("Current status was not start status, aborting. Status was: " .. string_from_Status(current_status))
+    return
+  end
+
+  local knowledge_description, content_type = knowledge_description()
+
+  if knowledge_description == last_knowledge_description then
+    return
+  end
+
+  if not knowledge_description then
+    debug_log('Poem description became nil, retrying...')
+    last_knowledge_description = nil
+    return
+  end
+
+  if content_type == Content_Type.unsupported then
+    gui_text = "This content type is not supported. Please select a " .. valid_content_type_list .. "."
+    last_knowledge_description = nil
+    return
+  end
+
+  debug_log('Got new ' .. content_type .. " description: " .. knowledge_description .. "\n")
+  last_knowledge_description = knowledge_description
+  gui_text = "Generating text from description, please wait...  "
+  debug_log("Submitting request to OpenAI remote service... \n")
+  send(request_from(knowledge_description, content_type))
+end
+
+--
+-- GUI: Overlay
+--
+
+GPTBannerOverlay = defclass(GPTBannerOverlay, overlay.OverlayWidget)
+GPTBannerOverlay.ATTRS{
+    default_pos={x=-35,y=-2},
+    default_enabled=true,
+    viewscreens={'dwarfmode/ViewSheets/UNIT','dwarfmode/ViewSheets/ITEM'},
+    frame={w=30, h=1},
+    frame_background=gui.CLEAR_PEN,
+}
+
+function GPTBannerOverlay:init()
+    self:addviews{
+      widgets.TextButton{
+        frame={t=0, l=0},
+        label='AI Generation View',
+        key='CUSTOM_CTRL_G',
+        on_activate=function() view = view and view:raise() or GPTScreen{}:show() end,
+      },
+    }
+end
+
+function GPTBannerOverlay:onInput(keys)
+  if GPTBannerOverlay.super.onInput(self, keys) then return true end
+
+  if keys._MOUSE_R_DOWN or keys.LEAVESCREEN then
+    if view then
+      view:dismiss()
+    end
+  end
+end
+
+OVERLAY_WIDGETS = {
+    gptbanner=GPTBannerOverlay,
+}
+
+--
+-- GUI: Window
+--
+
+local default_frame = {w=60, h=30, l=10, t=5}
+
+GPTWindow = defclass(GPTWindow, widgets.Window)
+GPTWindow.ATTRS{
+    frame_title='Generated Text',
+    resize_min=default_frame,
+    resizable=true,
+  }
+
+function GPTWindow:init()
+  self:addviews{
+    widgets.WrappedLabel{
+        view_id='label',
+        frame={t=0, l=0, r=0, b=0},
+        auto_height=false,
+        text_to_wrap=function() return gui_text end,
+        text_pen=COLOR_YELLOW,
+    },
+  }
+  self.frame=copyall(config.data.frame or default_frame)
+end
+
+function GPTWindow:onRenderFrame(dc, rect)
+    GPTWindow.super.onRenderFrame(self, dc, rect)
+
+    if current_status == Status.start then fetch_generated_text()
+    elseif current_status == Status.done then return
+    elseif client ~= nil then
+      if poll_count == polling_interval then
+        poll(client)
+        poll_count = 0
+      else
+        poll_count = poll_count + 1
+      end
+    end
+
+    self.subviews.label:updateLayout()
```

Comment:
this can get expensive. better to only update the layout when the text actually changes

Reply:
Maybe this is what's making the pcall() lock up the UI updating? I'll play around with it tomorrow.